README.md

```diff
@@ -188,3 +188,11 @@ pk crypto verify ./signed_file --detach-sig='./signed_file.sig'
 
 ## Verbosity
 TODO: explain verbosity levels when it is implemented
+
+# Build
+## Proto Files
+All `.proto` files are stored in the the `proto` directory. JavaScript and type definition files are build using the following command:
+
+```
+npm run build:proto
```

Comment:
Should mention how this is integrated into the webpack build system?

Reply:
This is a work in progress since `grpc_tools_node_protoc` doesn't seem to work on nixos, might need your help on that one, it should install with `grpc-tools` but `grpc_tools_node_protoc` doesn't seem to be added to the path or installed properly

Reply:
Does it require native compilation? Is the failure occurring during `npm install`? If so, there are ways around it...

Reply:
Some npm installs require compilation. This can be done in a nix-shell if the nix-shell has all the tooling required. There's quite a bit of stuff like `node-pre-gyp` or whatever. And python2. I did this before when I was trying to put `dat` and other binary npm stuff into Nixpkgs: https://github.com/svanderburg/node2nix/issues/71 and https://github.com/NixOS/nixpkgs/pull/32940.

However that's for using the packages with CLI executables in Nixpkgs itself. That may be the right idea to do if we can use the protoc compilers from `shell.nix` and that means pushing those packages to upstream (but first in our overlay).

Of course bringing all the tools needed to do compilation for Node is a bit of a mess.

Reply:
BTW have you seen this: https://github.com/agreatfool/grpc_tools_node_protoc_ts? It builds on top of the grpc_tools_node_protoc to generate typescript code instead of just JS code.

Reply:
There you see: https://github.com/grpc/grpc-node/blob/master/packages/grpc-tools/package.json#L27. The `grpc-tools` package requires using `node-pre-gyp`. 

As far as I remember, native Node dependencies has 3 kinds of buildchains:

* node-gyp-build
* node-gyp
* node-pre-gyp

So grpc-tools is using `node-pre-gyp`.

I wrote the section on native compilation here: https://github.com/MatrixAI/nixpkgs/blob/master/doc/languages-frameworks/node.section.md

So packaging that into Nix might be the right idea. You can start by putting all the build tools into your shell.nix.

Reply:
I have added in `nodePackages.node-pre-gyp` to `nativeBuildInputs` within `shell.nix`. I am getting a property missing error. In the section on native compilation, the addition is to `buildInputs` so maybe that is the issue?

Reply:
Funnily enough, npm seems to download `node-pre-gyp` and compile just fine until it gets to actually using the command, in which this error appears:
```
➜  js-polykey git:(ipc) ✗ grpc_tools_node_protoc                                                                             ~/Documents/github/js-polykey
/home/robbie/Documents/github/js-polykey/node_modules/grpc-tools/bin/protoc.js:41
    throw error;
    ^

Error: spawn /home/robbie/Documents/github/js-polykey/node_modules/grpc-tools/bin/protoc ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:264:19)
    at onErrorNT (internal/child_process.js:456:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  errno: 'ENOENT',
  code: 'ENOENT',
  syscall: 'spawn /home/robbie/Documents/github/js-polykey/node_modules/grpc-tools/bin/protoc',
  path: '/home/robbie/Documents/github/js-polykey/node_modules/grpc-tools/bin/protoc',
  spawnargs: [
    '--plugin=protoc-gen-grpc=/home/robbie/Documents/github/js-polykey/node_modules/grpc-tools/bin/grpc_node_plugin'
  ],
  cmd: '/home/robbie/Documents/github/js-polykey/node_modules/grpc-tools/bin/protoc --plugin=protoc-gen-grpc=/home/robbie/Documents/github/js-polykey/node_modules/grpc-tools/bin/grpc_node_plugin'
}
```

Reply:
I think this will be really useful, `grpc_tools_node_protoc` can compile the protos into js code that imports `grpc-js` directly and so we would have no need for loading protos anymore. So its worth getting this right.

Reply:
The node-pre-gyp would be best used as part of creating a new nodePackage in nixpkgs upstream. See the PR I created for dat. It's a similar process.

Reply:
ENOENT means something doesn't exist. Perhaps the tool is trying to execute something that doesn't exist. This can happen due to absolute paths or paths that don't expect a NixOS environment. I had a similar problem occur when trying to package dvc for Nixpkgs: https://github.com/iterative/dvc/issues/1509

You might want to chase down that error, and find out exaxtly what the path is failing.

Does this path exist? `/home/robbie/Documents/github/js-polykey/node_modules/grpc-tools/bin/protoc`.

Reply:
Recommend reading: https://github.com/svanderburg/node2nix

Reply:
And: https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/node.section.md

Reply:
I suggest putting up an issue in the grpc-node repo. I did this before for the `dvc` repo regarding "Packaging for NixOS". And put down your woes there. Someone there might know something about it and help. The `dvc` guys were helpful before too.